File: servers/flight/guard.py
"""Server-side JWT scope enforcement for Flight MCP (mirrors gateway/guard.ts)."""

from __future__ import annotations

import json
import logging
import os
import time
from dataclasses import dataclass, field
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

import jwt
import yaml

logger = logging.getLogger("mcp-tool-guard")

SERVER_ID = "flight"


@dataclass
class ToolConfig:
    required_scope: str
    alert: bool = False
    log_level: str = "info"


@dataclass
class GuardResult:
    allowed: bool
    required_scope: str
    reason: str | None = None
    token_scopes: list[str] = field(default_factory=list)


@dataclass
class AuditEntry:
    timestamp: str
    decision: str
    server: str
    tool: str
    required_scope: str
    token_scopes: list[str]
    reason: str | None = None
    alert: bool = False
    duration_ms: int | None = None


class FlightToolGuard:
    def __init__(self, tools: dict[str, ToolConfig], public_key_pem: str) -> None:
        self.tools = tools
        self.public_key_pem = public_key_pem
        self.enabled = os.environ.get("MCP_GUARD_ENABLED", "true").lower() != "false"
        self._audit: list[AuditEntry] = []

    @classmethod
    def load(cls) -> FlightToolGuard:
        config_path = Path(__file__).parent / "guard_config.yaml"
        raw = yaml.safe_load(config_path.read_text())
        tools_raw = raw.get("tools", {})
        tools = {
            name: ToolConfig(
                required_scope=cfg["required_scope"],
                alert=bool(cfg.get("alert", False)),
                log_level=str(cfg.get("log_level", "info")),
            )
            for name, cfg in tools_raw.items()
        }
        return cls(tools=tools, public_key_pem=_load_public_key_pem())

    @staticmethod
    def extract_scopes(payload: dict[str, Any]) -> list[str]:
        raw = payload.get("scope") or payload.get("scopes") or payload.get("scp")
        if not raw:
            return []
        if isinstance(raw, list):
            return [str(s) for s in raw]
        return [s for s in str(raw).replace(",", " ").split() if s]

    @staticmethod
    def has_scope(token_scopes: list[str], required: str) -> bool:
        if required in token_scopes:
            return True
        resource = required.split(":", 1)[0]
        return f"{resource}:*" in token_scopes or "*" in token_scopes

    def _log(self, entry: AuditEntry, tool_cfg: ToolConfig | None) -> None:
        self._audit.append(entry)
        payload = {
            "timestamp": entry.timestamp,
            "decision": entry.decision,
            "server": entry.server,
            "tool": entry.tool,
            "required_scope": entry.required_scope,
            "token_scopes": entry.token_scopes,
            "reason": entry.reason,
            "duration_ms": entry.duration_ms,
        }
        if tool_cfg and tool_cfg.alert:
            logger.warning("[MCPToolGuard ALERT] %s", json.dumps(payload))
        else:
            logger.info("[MCPToolGuard] %s", json.dumps(payload))

    def authorize(self, tool: str, bearer_token: str | None) -> GuardResult:
        start = time.perf_counter()
        tool_cfg = self.tools.get(tool)
        required = tool_cfg.required_scope if tool_cfg else "(unknown)"

        if not bearer_token:
            self._log(
                AuditEntry(
                    timestamp=_now(),
                    decision="deny",
                    server=SERVER_ID,
                    tool=tool,
                    required_scope=required,
                    token_scopes=[],
                    reason="Missing Authorization: Bearer token",
                    duration_ms=_ms(start),
                ),
                tool_cfg,
            )
            return GuardResult(
                allowed=False,
                required_scope=required,
                reason="Missing Authorization: Bearer token",
            )

        try:
            payload = jwt.decode(
                bearer_token,
                self.public_key_pem,
                algorithms=["RS256"],
            )
            scopes = self.extract_scopes(payload)
        except jwt.PyJWTError as exc:
            self._log(
                AuditEntry(
                    timestamp=_now(),
                    decision="deny",
                    server=SERVER_ID,
                    tool=tool,
                    required_scope=required,
                    token_scopes=[],
                    reason=f"JWT validation failed: {exc}",
                    duration_ms=_ms(start),
                ),
                tool_cfg,
            )
            return GuardResult(
                allowed=False,
                required_scope=required,
                reason=f"JWT validation failed: {exc}",
            )

        if not tool_cfg:
            reason = f"Tool '{tool}' not configured for server '{SERVER_ID}'"
            self._log(
                AuditEntry(
                    timestamp=_now(),
                    decision="deny",
                    server=SERVER_ID,
                    tool=tool,
                    required_scope="(unknown)",
                    token_scopes=scopes,
                    reason=reason,
                    duration_ms=_ms(start),
                ),
                None,
            )
            return GuardResult(allowed=False, required_scope="(unknown)", reason=reason, token_scopes=scopes)

        allowed = self.has_scope(scopes, tool_cfg.required_scope)
        reason = None if allowed else f"Missing required scope '{tool_cfg.required_scope}'"
        self._log(
            AuditEntry(
                timestamp=_now(),
                decision="allow" if allowed else "deny",
                server=SERVER_ID,
                tool=tool,
                required_scope=tool_cfg.required_scope,
                token_scopes=scopes,
                reason=reason,
                alert=tool_cfg.alert,
                duration_ms=_ms(start),
            ),
            tool_cfg,
        )
        return GuardResult(
            allowed=allowed,
            required_scope=tool_cfg.required_scope,
            reason=reason,
            token_scopes=scopes,
        )

    def recent_audit(self, limit: int = 100) -> list[dict[str, Any]]:
        return [
            {
                "timestamp": e.timestamp,
                "decision": e.decision,
                "server": e.server,
                "tool": e.tool,
                "required_scope": e.required_scope,
                "token_scopes": e.token_scopes,
                "reason": e.reason,
            }
            for e in self._audit[-limit:]
        ]


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


def _ms(start: float) -> int:
    return round((time.perf_counter() - start) * 1000)


def _load_public_key_pem() -> str:
    inline = os.environ.get("MCP_GUARD_PUBLIC_KEY_PEM")
    if inline:
        return inline.replace("\\n", "\n")

    root = Path(__file__).resolve().parent
    repo = root.parent.parent
    for candidate in (
        root / "demo-public.pem",
        repo / "ui" / "public" / "demo-public.pem",
        repo / "keys" / "demo-public.pem",
    ):
        if candidate.is_file():
            return candidate.read_text()

    raise RuntimeError(
        "No JWT public key found. Set MCP_GUARD_PUBLIC_KEY_PEM or run make keys."
    )

File: gateway/guard.ts
import {
  createRemoteJWKSet,
  decodeJwt,
  importSPKI,
  jwtVerify,
} from "jose";
import { parse as parseYaml } from "yaml";

import { AuditLogger } from "./logger.js";
import type {
  AuditContext,
  AuditLogEntry,
  GuardConfig,
  GuardResult,
  JwtPayload,
  ServerConfig,
  ToolConfig,
} from "./types.js";

/**
 * Validates a bearer token and derives scopes from it. Implementations own
 * all trust-decision logic (which key/JWKS to verify against, issuer/audience
 * checks, claims shape, etc). ToolGuard only consumes this interface — it
 * never re-implements validation itself.
 *
 * Future IdP-specific implementations (e.g. a Keycloak or Entra validator)
 * are expected to satisfy this same interface so ToolGuard doesn't need to
 * change when a new IdP is added.
 */
export interface JwtValidator {
  /** Verify signature/claims and return the payload plus derived scopes. Throws on invalid tokens. */
  validateToken(token: string): Promise<{ payload: JwtPayload; scopes: string[] }>;
  /** Derive a normalized scope list from an already-verified payload. */
  extractScopes(payload: JwtPayload): string[];
  /** Check whether `tokenScopes` satisfies `required` (supports `resource:*` and `*` wildcards). */
  hasScope(tokenScopes: string[], required: string): boolean;
  /** Optional async setup (e.g. importing a PEM key) to run before first use. */
  init?(): Promise<void>;
}

export interface JwtValidatorOptions {
  publicKey?: CryptoKey | string;
  algorithm?: string;
  /** Optional server-side callback to enforce immediate M2M revocation after agent delete. */
  isM2mClientActive?: (clientId: string) => Promise<boolean>;
  /** IdP issuer — when token `iss` matches, verify via JWKS instead of PEM. */
  jwtIssuer?: string;
  jwtAudience?: string;
  jwksUrl?: string;
}

/**
 * Default JwtValidator: dual-trust PEM-or-JWKS verification.
 *
 * - If the token's `iss` claim matches the configured `jwtIssuer` (and a JWKS
 *   URL/audience are configured), verify via JWKS and additionally enforce
 *   M2M-agent liveness (`assertActiveM2mAgent`) for machine-to-machine tokens.
 * - Otherwise, fall back to PEM-based verification.
 */
export class DefaultJwtValidator implements JwtValidator {
  private publicKey?: CryptoKey;
  private algorithm: string;
  private jwtIssuer?: string;
  private jwtAudience?: string;
  private jwks?: ReturnType<typeof createRemoteJWKSet>;
  private isM2mClientActive?: (clientId: string) => Promise<boolean>;
  private publicKeyPromise?: Promise<CryptoKey>;

  constructor(options: JwtValidatorOptions) {
    this.algorithm = options.algorithm ?? "RS256";
    this.isM2mClientActive = options.isM2mClientActive;
    this.jwtIssuer = options.jwtIssuer?.replace(/\/$/, "");
    this.jwtAudience = options.jwtAudience;
    if (options.jwksUrl) {
      this.jwks = createRemoteJWKSet(new URL(options.jwksUrl));
    }
    if (options.publicKey && typeof options.publicKey === "string") {
      this.publicKeyPromise = importSPKI(options.publicKey, this.algorithm);
    } else if (options.publicKey && typeof options.publicKey !== "string") {
      this.publicKey = options.publicKey;
    }
  }

  async init(): Promise<void> {
    if (this.publicKeyPromise) {
      this.publicKey = await this.publicKeyPromise;
    }
  }

  extractScopes(payload: JwtPayload): string[] {
    const scopes: string[] = [];
    const raw = payload.scope ?? payload.scopes ?? payload.scp;
    if (raw) {
      if (Array.isArray(raw)) scopes.push(...raw.map(String));
      else scopes.push(...String(raw).split(/[\s,]+/).filter(Boolean));
    }
    if (Array.isArray(payload.permissions)) {
      scopes.push(...payload.permissions.map(String));
    }
    return [...new Set(scopes)];
  }

  hasScope(tokenScopes: string[], required: string): boolean {
    if (tokenScopes.includes(required)) return true;
    const [resource] = required.split(":");
    return tokenScopes.includes(`${resource}:*`) || tokenScopes.includes("*");
  }

  private issMatches(tokenIss: unknown): boolean {
    if (!this.jwtIssuer || typeof tokenIss !== "string") return false;
    return tokenIss.replace(/\/$/, "") === this.jwtIssuer;
  }

  private isM2mLikeToken(payload: JwtPayload): boolean {
    // Primary signal: Auth0 M2M token subject/client-id shape.
    // Secondary signal: explicit grant type claim when present.
    return this.clientIdFromPayload(payload) !== null || payload.gty === "client-credentials";
  }

  private clientIdFromPayload(payload: JwtPayload): string | null {
    if (typeof payload.client_id === "string" && payload.client_id.trim()) {
      return payload.client_id.trim();
    }
    if (typeof payload.sub === "string") {
      const match = payload.sub.match(/^([^@]+)@clients$/);
      if (match?.[1]) return match[1];
    }
    return null;
  }

  private async assertActiveM2mAgent(payload: JwtPayload): Promise<void> {
    if (!this.isM2mLikeToken(payload)) return;
    const clientId = this.clientIdFromPayload(payload);
    if (!clientId) {
      throw new Error("M2M token missing client_id/sub claim shape");
    }
    if (!this.isM2mClientActive) return;
    const active = await this.isM2mClientActive(clientId);
    if (!active) {
      throw new Error("Agent revoked or deleted");
    }
  }

  async validateToken(token: string): Promise<{ payload: JwtPayload; scopes: string[] }> {
    const unverified = decodeJwt(token) as JwtPayload;
    if (
      this.jwks &&
      this.jwtIssuer &&
      this.jwtAudience &&
      this.issMatches(unverified.iss)
    ) {
      const { payload } = await jwtVerify(token, this.jwks, {
        issuer: `${this.jwtIssuer}/`,
        audience: this.jwtAudience,
      });
      const jwtPayload = payload as JwtPayload;
      await this.assertActiveM2mAgent(jwtPayload);
      return { payload: jwtPayload, scopes: this.extractScopes(jwtPayload) };
    }

    if (!this.publicKey) {
      throw new Error("Public key not configured — call init() with a PEM key");
    }
    const { payload } = await jwtVerify(token, this.publicKey, {
      algorithms: [this.algorithm],
    });
    const jwtPayload = payload as JwtPayload;
    return { payload: jwtPayload, scopes: this.extractScopes(jwtPayload) };
  }
}

export interface ToolGuardOptions extends JwtValidatorOptions {
  config: GuardConfig | string;
  logger?: AuditLogger;
  /** Inject a custom JwtValidator (e.g. for a different IdP). Defaults to a DefaultJwtValidator built from the PEM/JWKS options above. */
  jwtValidator?: JwtValidator;
}

export class ToolGuard {
  private config: GuardConfig;
  readonly logger: AuditLogger;
  private validator: JwtValidator;

  constructor(options: ToolGuardOptions) {
    this.config =
      typeof options.config === "string"
        ? (parseYaml(options.config) as GuardConfig)
        : options.config;
    this.logger = options.logger ?? new AuditLogger();
    this.validator =
      options.jwtValidator ??
      new DefaultJwtValidator({
        publicKey: options.publicKey,
        algorithm: options.algorithm,
        isM2mClientActive: options.isM2mClientActive,
        jwtIssuer: options.jwtIssuer,
        jwtAudience: options.jwtAudience,
        jwksUrl: options.jwksUrl,
      });
  }

  async init(): Promise<void> {
    await this.validator.init?.();
  }

  /** The injected (or default) JwtValidator — exposed for call sites that need to validate/check scopes outside of authorize(). */
  get jwtValidator(): JwtValidator {
    return this.validator;
  }

  getServerConfig(server: string): ServerConfig | undefined {
    return this.config.servers[server];
  }

  getToolConfig(server: string, tool: string): ToolConfig | undefined {
    return this.config.servers[server]?.tools[tool];
  }

  listServers(): string[] {
    return Object.keys(this.config.servers);
  }

  listTools(server: string): string[] {
    const cfg = this.config.servers[server];
    return cfg ? Object.keys(cfg.tools) : [];
  }

  /** Replace in-memory policy (runtime registry updates). */
  replaceConfig(config: GuardConfig): void {
    this.config = config;
  }

  checkScope(
    server: string,
    tool: string,
    tokenScopes: string[],
    audit?: AuditContext,
  ): GuardResult {
    const toolConfig = this.getToolConfig(server, tool);
    const timestamp = new Date().toISOString();

    if (!toolConfig) {
      const entry: AuditLogEntry = {
        timestamp,
        decision: "deny",
        server,
        tool,
        required_scope: "(unknown)",
        token_scopes: tokenScopes,
        reason: `Tool '${tool}' not configured for server '${server}'`,
        ...audit,
        source: audit?.source ?? "proxy",
      };
      this.logger.log(entry);
      return { allowed: false, reason: entry.reason, required_scope: "(unknown)", entry };
    }

    const required = toolConfig.required_scope;
    const allowed = this.validator.hasScope(tokenScopes, required);

    const entry: AuditLogEntry = {
      timestamp,
      decision: allowed ? "allow" : "deny",
      server,
      tool,
      required_scope: required,
      token_scopes: tokenScopes,
      alert: toolConfig.alert,
      log_level: toolConfig.log_level,
      reason: allowed
        ? undefined
        : `Missing required scope '${required}'`,
      ...audit,
      source: audit?.source ?? "proxy",
    };

    this.logger.log(entry);
    return {
      allowed,
      reason: entry.reason,
      required_scope: required,
      entry,
    };
  }

  async authorize(
    server: string,
    tool: string,
    token: string,
    audit?: AuditContext,
  ): Promise<GuardResult> {
    const start = performance.now();
    try {
      const { scopes } = await this.validator.validateToken(token);
      const result = this.checkScope(server, tool, scopes, audit);
      result.entry.duration_ms = Math.round(performance.now() - start);
      return result;
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      const entry: AuditLogEntry = {
        timestamp: new Date().toISOString(),
        decision: "deny",
        server,
        tool,
        required_scope: this.getToolConfig(server, tool)?.required_scope ?? "(unknown)",
        token_scopes: [],
        reason: `JWT validation failed: ${message}`,
        duration_ms: Math.round(performance.now() - start),
        ...audit,
        source: audit?.source ?? "proxy",
      };
      this.logger.log(entry);
      return {
        allowed: false,
        reason: entry.reason,
        required_scope: entry.required_scope,
        entry,
      };
    }
  }
}
